numpy/testing/utils.py

```diff
@@ -343,16 +343,29 @@ def assert_equal(actual,desired,err_msg='',verbose=True):
         except AssertionError:
             raise AssertionError(msg)
 
+    def isnat(x):
+        return (hasattr(x, 'dtype')
+                and getattr(x.dtype, 'kind', '_') in 'mM'
+                and x != x)
+
     # Inf/nan/negative zero handling
     try:
         # isscalar test to check cases such as [np.nan] != np.nan
-        if isscalar(desired) != isscalar(actual):
+        if (isscalar(desired) != isscalar(actual)
+                and not (isinstance(desired, dtype)
```

Comment:
There are some masked array specific tests that use assert_equals to compare a dtype and the string version of a dtype (e.g., "float"). One of these is a scalar, the other is not.

On Wed, Jan 13, 2016 at 3:57 PM, Charles Harris <notifications@github.com>
wrote:

>>      # Inf/nan/negative zero handling
>>      try:
>>          # isscalar test to check cases such as [np.nan] != np.nan
>> -        if isscalar(desired) != isscalar(actual):
>> +        if (isscalar(desired) != isscalar(actual)
>> +                and not (isinstance(desired, dtype)
> Bit curious as to why `dtype` needs to be checked.
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/numpy/numpy/pull/7001/files#r49670305